xds/src/test/java/io/grpc/xds/LookasideChannelLbWithLocalityStoreTest.java

```diff
@@ -0,0 +1,297 @@
+  @Mock
+  private XdsCommsFactory xdsCommsFactory;
+  @Mock
+  private PickSubchannelArgs pickSubchannelArgs;
+  @Mock
+  private ManagedChannel channel;
+  @Mock
+  private AbstractXdsComms xdsComms;
+
+  private final SynchronizationContext syncContext = new SynchronizationContext(
+      new Thread.UncaughtExceptionHandler() {
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+          throw new AssertionError(e);
+        }
+      });
+
+  private final LoadBalancerRegistry lbRegistry = new LoadBalancerRegistry();
+  private final Map<String, LoadBalancer> loadBalancers = new HashMap<>();
+  private final Map<String, Helper> childHelpers = new HashMap<>();
+
+  private final LocalityStoreFactory localityStoreFactory = new LocalityStoreFactoryImpl();
+
+  private final LoadBalancerProvider lbProvider = new LoadBalancerProvider() {
+
+    @Override
+    public boolean isAvailable() {
+      return true;
+    }
+
+    @Override
+    public int getPriority() {
+      return 0;
+    }
+
+    @Override
+    public String getPolicyName() {
+      return "round_robin";
+    }
+
+    @Override
+    public LoadBalancer newLoadBalancer(Helper helper) {
+      LoadBalancer fakeLb = mock(LoadBalancer.class);
+      loadBalancers.put(helper.getAuthority(), fakeLb);
+      childHelpers.put(helper.getAuthority(), helper);
+      return fakeLb;
+    }
+  };
+
+  private AdsStreamCallback2 adsStreamCallback2;
+  private LookasideChannelLb lookasideChannelLb;
+
+  @Before
+  public void setUp() {
+    lbRegistry.register(lbProvider);
+
+    doReturn(mock(ChannelLogger.class)).when(helper).getChannelLogger();
+    doReturn(syncContext).when(helper).getSynchronizationContext();
+    doReturn(channel).when(helper).createResolvingOobChannel(BALANCER_NAME);
+    doReturn(lrsClient).when(lrsClientFactory).createLoadReportClient(
+        same(channel), same(helper), any(BackoffPolicy.Provider.class),
+        any(LoadStatsStore.class));
+    doReturn(xdsComms).when(xdsCommsFactory).newXdsComms(
+        same(channel), any(AdsStreamCallback2.class));
+
+    lookasideChannelLb = new LookasideChannelLb(
+        helper, adsStreamCallback, BALANCER_NAME, lrsClientFactory, lbRegistry,
+        localityStoreFactory, xdsCommsFactory);
+
+    verify(helper).createResolvingOobChannel(BALANCER_NAME);
+    verify(lrsClientFactory).createLoadReportClient(
+        same(channel), same(helper), isA(BackoffPolicy.Provider.class), any(LoadStatsStore.class));
+    ArgumentCaptor<AdsStreamCallback2> adsStreamCallback2Captor =
+        ArgumentCaptor.forClass(AdsStreamCallback2.class);
+    verify(xdsCommsFactory).newXdsComms(same(channel), adsStreamCallback2Captor.capture());
+    adsStreamCallback2 = adsStreamCallback2Captor.getValue();
+    verify(xdsComms).start();
+  }
+
+  @After
+  public void tearDown() {
+    // release timers
+    lookasideChannelLb.shutdown();
+  }
+
+  @Test
+  public void verifyUpdateLocalityStore() {
```

Comment:
Isn't this a pure `LocalityStore`'s behavior test? This should go into `LocalityStoreTest`. For testing interaction between `LookasideChannelLb` and `LocalityStore`, you've already had https://github.com/grpc/grpc-java/pull/6093/files#diff-bf39fa0a326861e0b7f8cc061b9adbf7R228. This test class may not be necessary.

Reply:
That one you referenced is a mock LocalityStore, which just verifies some of its methods get called. Can not verify helper.updateBalancingState, and picked pickers. This one is kinda end to end, and verifies the right helper gets updated.

Reply:
Yes, a mock `LocalityStore` is enough for verifying the interactions between `LookasideChannelLb` and `LocalityStore`. Verifying `helper.updateBalancingState` propagation and picked pickers can be done in `LocalityStoreTest`.

I mean we should either have `LookasideChannelLb` + `LocalityStore` tested as a whole, with a RPC service, for the the flow of 
- establishing communication to traffic director and receive EDS response
- `LookasideChannelLb` delivers load balancing responsibility to `LocalityStore`
- `LocalityStore` performs load balancing until subchannels are picked

Or have them test separately (a with a service for `LookasideChannelLb`'s test and crafted EDS data for `LocalityStore`'s test): a mocked `LocalityStore` is the point for `LookasideChannelLb` interacting with `LocalityStore` and a mocked `Helper` is the point for `LocalityStore` propagates balancing state to `LookasideChannelLb`.

Having two places partially tested doesn't cover the whole logic and redundantly verifies subchannel picking behavior of `LocalityStore again and again.

Reply:
Removed this test.